File: packages/node/src/client/methods/topics/topics.ts
/**
 * Forum topic management methods for Bot API.
 *
 * @packageDocumentation
 */

import { BotProfileMethods } from "./profile.js";
import type { ForumTopic } from "../../types/index.js";

/**
 * Mixin providing forum topic creation, editing, closing, and pin operations.
 */
export abstract class ForumTopicMethods extends BotProfileMethods {
  /**
   * Creates a topic in a forum supergroup chat.
   *
   * @param options - Options including target `chat_id`, `name`, and optional `icon_color` or `icon_custom_emoji_id`.
   * @returns The created {@link ForumTopic} object on success.
   * @throws {@link TelegramApiError} When creating topic fails.
   *
   * @example
   * ```ts
   * const topic = await bot.createForumTopic({
   *   chat_id: -1001234567890,
   *   name: "Announcements",
   *   icon_color: 0x6FB9F0,
   * });
   * console.log(`Created thread ID: ${topic.message_thread_id}`);
   * ```
   */
  public async createForumTopic(options: {
    chat_id: number | string;
    name: string;
    icon_color?: number;
    icon_custom_emoji_id?: string;
  }): Promise<ForumTopic> {
    return this.request<ForumTopic>(
      "createForumTopic",
      options as unknown as Record<string, unknown>,
    );
  }

  /**
   * Edits the name and icon of a topic in a forum supergroup chat.
   *
   * @param options - Options including `chat_id`, `message_thread_id`, and updated `name` or `icon_custom_emoji_id`.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When editing topic fails.
   *
   * @example
   * ```ts
   * await bot.editForumTopic({
   *   chat_id: -1001234567890,
   *   message_thread_id: 42,
   *   name: "Updated Announcements",
   * });
   * ```
   */
  public async editForumTopic(options: {
    chat_id: number | string;
    message_thread_id: number;
    name?: string;
    icon_custom_emoji_id?: string;
  }): Promise<boolean> {
    return this.request<boolean>("editForumTopic", options as unknown as Record<string, unknown>);
  }

  /**
   * Closes an open topic in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @param messageThreadId - Unique identifier for the target message thread of the forum topic.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When closing topic fails.
   *
   * @example
   * ```ts
   * await bot.closeForumTopic(-1001234567890, 42);
   * ```
   */
  public async closeForumTopic(chatId: number | string, messageThreadId: number): Promise<boolean> {
    return this.request<boolean>("closeForumTopic", {
      chat_id: chatId,
      message_thread_id: messageThreadId,
    });
  }

  /**
   * Reopens a closed topic in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @param messageThreadId - Unique identifier for the target message thread of the forum topic.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When reopening topic fails.
   *
   * @example
   * ```ts
   * await bot.reopenForumTopic(-1001234567890, 42);
   * ```
   */
  public async reopenForumTopic(
    chatId: number | string,
    messageThreadId: number,
  ): Promise<boolean> {
    return this.request<boolean>("reopenForumTopic", {
      chat_id: chatId,
      message_thread_id: messageThreadId,
    });
  }

  /**
   * Deletes a forum topic along with all its messages in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @param messageThreadId - Unique identifier for the target message thread of the forum topic.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When deleting topic fails.
   *
   * @example
   * ```ts
   * await bot.deleteForumTopic(-1001234567890, 42);
   * ```
   */
  public async deleteForumTopic(
    chatId: number | string,
    messageThreadId: number,
  ): Promise<boolean> {
    return this.request<boolean>("deleteForumTopic", {
      chat_id: chatId,
      message_thread_id: messageThreadId,
    });
  }

  /**
   * Clears the list of pinned messages in a forum topic.
   *
   * @param chatId - Unique identifier for the target chat.
   * @param messageThreadId - Unique identifier for the target message thread of the forum topic.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When unpinning messages fails.
   *
   * @example
   * ```ts
   * await bot.unpinAllForumTopicMessages(-1001234567890, 42);
   * ```
   */
  public async unpinAllForumTopicMessages(
    chatId: number | string,
    messageThreadId: number,
  ): Promise<boolean> {
    return this.request<boolean>("unpinAllForumTopicMessages", {
      chat_id: chatId,
      message_thread_id: messageThreadId,
    });
  }

  /**
   * Edits the name of the 'General' topic in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @param name - New name of the topic (1-128 characters).
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When editing General topic fails.
   *
   * @example
   * ```ts
   * await bot.editGeneralForumTopic(-1001234567890, "General Chat");
   * ```
   */
  public async editGeneralForumTopic(chatId: number | string, name: string): Promise<boolean> {
    return this.request<boolean>("editGeneralForumTopic", { chat_id: chatId, name });
  }

  /**
   * Closes an open 'General' topic in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When closing General topic fails.
   *
   * @example
   * ```ts
   * await bot.closeGeneralForumTopic(-1001234567890);
   * ```
   */
  public async closeGeneralForumTopic(chatId: number | string): Promise<boolean> {
    return this.request<boolean>("closeGeneralForumTopic", { chat_id: chatId });
  }

  /**
   * Reopens a closed 'General' topic in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When reopening General topic fails.
   *
   * @example
   * ```ts
   * await bot.reopenGeneralForumTopic(-1001234567890);
   * ```
   */
  public async reopenGeneralForumTopic(chatId: number | string): Promise<boolean> {
    return this.request<boolean>("reopenGeneralForumTopic", { chat_id: chatId });
  }

  /**
   * Hides the 'General' topic in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When hiding General topic fails.
   *
   * @example
   * ```ts
   * await bot.hideGeneralForumTopic(-1001234567890);
   * ```
   */
  public async hideGeneralForumTopic(chatId: number | string): Promise<boolean> {
    return this.request<boolean>("hideGeneralForumTopic", { chat_id: chatId });
  }

  /**
   * Unhides the 'General' topic in a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When unhiding General topic fails.
   *
   * @example
   * ```ts
   * await bot.unhideGeneralForumTopic(-1001234567890);
   * ```
   */
  public async unhideGeneralForumTopic(chatId: number | string): Promise<boolean> {
    return this.request<boolean>("unhideGeneralForumTopic", { chat_id: chatId });
  }

  /**
   * Clears the list of pinned messages in the 'General' topic of a forum supergroup chat.
   *
   * @param chatId - Unique identifier for the target chat.
   * @returns `true` on success.
   * @throws {@link TelegramApiError} When unpinning General topic messages fails.
   *
   * @example
   * ```ts
   * await bot.unpinAllGeneralForumTopicMessages(-1001234567890);
   * ```
   */
  public async unpinAllGeneralForumTopicMessages(chatId: number | string): Promise<boolean> {
    return this.request<boolean>("unpinAllGeneralForumTopicMessages", { chat_id: chatId });
  }
}

File: packages/python/src/telebot_py/bot/topics.py
"""Forum topic Bot API methods (parity with packages/go/pkg/bot/topics.go)."""

from __future__ import annotations

from telebot_py.bot.base import Requester, clean_payload, parse_flag, parse_result
from telebot_py.types.topics import ForumTopic


class TopicsMixin(Requester):
    """Bot methods for creating and closing forum topics."""

    async def create_forum_topic(
        self,
        chat_id: int | str,
        name: str,
        *,
        icon_color: int | None = None,
        icon_custom_emoji_id: str | None = None,
    ) -> ForumTopic:
        """Create a topic in a forum supergroup chat.

        The bot must have the ``can_manage_topics`` administrator right.

        Example:
            >>> topic = await bot.create_forum_topic(-100, "General")

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            name: Topic name; 1-128 characters.
            icon_color: Color of the topic icon in RGB format.
            icon_custom_emoji_id: Unique identifier of the custom emoji shown
                as the topic icon.

        Returns:
            The created ForumTopic.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            name=name,
            icon_color=icon_color,
            icon_custom_emoji_id=icon_custom_emoji_id,
        )
        return parse_result(ForumTopic, await self.request("createForumTopic", payload))

    async def edit_forum_topic(
        self,
        chat_id: int | str,
        message_thread_id: int,
        *,
        name: str | None = None,
        icon_custom_emoji_id: str | None = None,
    ) -> bool:
        """Edit the name and icon of a topic in a forum supergroup chat.

        The bot must have the ``can_manage_topics`` administrator right, unless
        it is the creator of the topic.

        Example:
            >>> ok = await bot.edit_forum_topic(-100, 4, name="Announcements")

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            message_thread_id: Unique identifier of the target message thread
                of the forum topic; must be positive.
            name: New topic name; 1-128 characters.
            icon_custom_emoji_id: New unique identifier of the custom emoji
                shown as the topic icon.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(
            chat_id=chat_id,
            message_thread_id=message_thread_id,
            name=name,
            icon_custom_emoji_id=icon_custom_emoji_id,
        )
        return parse_flag(await self.request("editForumTopic", payload))

    async def close_forum_topic(self, chat_id: int | str, message_thread_id: int) -> bool:
        """Close an open topic in a forum supergroup chat.

        Example:
            >>> ok = await bot.close_forum_topic(-100, 4)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            message_thread_id: Unique identifier of the target message thread.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, message_thread_id=message_thread_id)
        return parse_flag(await self.request("closeForumTopic", payload))

    async def reopen_forum_topic(self, chat_id: int | str, message_thread_id: int) -> bool:
        """Reopen a closed topic in a forum supergroup chat.

        The bot must have the ``can_manage_topics`` administrator right, unless
        it is the creator of the topic.

        Example:
            >>> ok = await bot.reopen_forum_topic(-100, 4)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            message_thread_id: Unique identifier of the target message thread
                of the forum topic.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, message_thread_id=message_thread_id)
        return parse_flag(await self.request("reopenForumTopic", payload))

    async def delete_forum_topic(self, chat_id: int | str, message_thread_id: int) -> bool:
        """Delete a forum topic along with all its messages.

        The bot must have the ``can_manage_topics`` administrator right, unless
        it is the creator of the topic. Implies closing the topic first.

        Example:
            >>> ok = await bot.delete_forum_topic(-100, 4)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            message_thread_id: Unique identifier of the target message thread
                of the forum topic.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, message_thread_id=message_thread_id)
        return parse_flag(await self.request("deleteForumTopic", payload))

    async def unpin_all_forum_topic_messages(
        self, chat_id: int | str, message_thread_id: int
    ) -> bool:
        """Clear the list of pinned messages in a forum topic.

        The bot must have the ``can_pin_messages`` administrator right in the
        topic to do this.

        Example:
            >>> ok = await bot.unpin_all_forum_topic_messages(-100, 4)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            message_thread_id: Unique identifier of the target message thread
                of the forum topic.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, message_thread_id=message_thread_id)
        return parse_flag(await self.request("unpinAllForumTopicMessages", payload))

    async def edit_general_forum_topic(self, chat_id: int | str, name: str) -> bool:
        """Edit the name of the 'General' topic in a forum supergroup chat.

        The bot must have the ``can_manage_topics`` administrator right. The
        name of the 'General' topic can never be empty.

        Example:
            >>> ok = await bot.edit_general_forum_topic(-100, "General Chat")

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.
            name: New topic name; 1-128 characters.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id, name=name)
        return parse_flag(await self.request("editGeneralForumTopic", payload))

    async def close_general_forum_topic(self, chat_id: int | str) -> bool:
        """Close an open 'General' topic in a forum supergroup chat.

        The bot must have the ``can_manage_topics`` administrator right.

        Example:
            >>> ok = await bot.close_general_forum_topic(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("closeGeneralForumTopic", payload))

    async def reopen_general_forum_topic(self, chat_id: int | str) -> bool:
        """Reopen a closed 'General' topic in a forum supergroup chat.

        The bot must have the ``can_manage_topics`` administrator right.

        Example:
            >>> ok = await bot.reopen_general_forum_topic(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("reopenGeneralForumTopic", payload))

    async def hide_general_forum_topic(self, chat_id: int | str) -> bool:
        """Hide the 'General' topic in a forum supergroup chat.

        The bot must be a member of the group with the ``can_manage_topics``
        administrator right. The 'General' topic is automatically unhidden when
        the chat is opened by other members.

        Example:
            >>> ok = await bot.hide_general_forum_topic(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("hideGeneralForumTopic", payload))

    async def unhide_general_forum_topic(self, chat_id: int | str) -> bool:
        """Unhide the 'General' topic in a forum supergroup chat.

        The bot must have the ``can_manage_topics`` administrator right.

        Example:
            >>> ok = await bot.unhide_general_forum_topic(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("unhideGeneralForumTopic", payload))

    async def unpin_all_general_forum_topic_messages(self, chat_id: int | str) -> bool:
        """Clear the list of pinned messages in a 'General' forum topic.

        The bot must have the ``can_pin_messages`` administrator right in the
        supergroup to do this.

        Example:
            >>> ok = await bot.unpin_all_general_forum_topic_messages(-100)

        Args:
            chat_id: Unique identifier for the target chat or username of the
                target supergroup.

        Returns:
            True on success.

        Raises:
            InvalidTokenError: If Telegram rejects the token (HTTP 401).
            TelegramApiError: If Telegram responds not-ok or retries exhaust.
            NetworkError: If the transport keeps failing after retries.
        """
        payload = clean_payload(chat_id=chat_id)
        return parse_flag(await self.request("unpinAllGeneralForumTopicMessages", payload))
